In [123]:
import pandas as pd
from random import *
import numpy as np

In [124]:
df= pd.read_csv('ds_salaries_real01.csv',index_col='id')
df

work_year experience_level employment_type                   job_title  \
id                                                                            
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary salary_currency employee_residence remote_ratio company_location  \
id                                                                             
0     70000             EUR                 DE            0               DE   
1    260000             USD                 JP            0               JP   
2     85000             GBP                 GB           50               GB   
3    -20000             USD                 HN            0               HN   
4    150000             USD                 US          ???               US   
..      ...             ...                ...          ...              ...   
602  154000             USD                 US          100               US   
603  126000             USD                 US          100               US   
604  129000             USD                 US            0               US   
605  150000             USD                 US          100               US   
606 -200000             USD                 IN          100               US   

    company_size  
id                
0              L  
1              S  
2              M  
3              S  
4              L  
..           ...  
602            M  
603            M  
604            M  
605            M  
606            L  

[607 rows x 10 columns]

# Schritt 2 Datenreinigung

<div class="alert alert-block alert-info"><b>Szenario</b>

Wir wollen die den Gehaltsdurchschnitt eines Daten Analysten ermitteln.
Leider ist und die Katze über die Tastatur gelaufen und hat unser Daten verunreinigt.
Wir müssen sie im folgenden Abschnitt reinigen.
- 30 Werte in der Spalte 'salary' sind betroffen
- 30 Werte ind er Spalte 'remote_ratio' sind betroffen

</div>

## *Fragen an das Publikum* :
### Berechne den durchschnitt der ersten 5 Gehälter von hand.
**Was könnte hir ein Problem darstellen**

In [125]:
df.head()

work_year experience_level employment_type                   job_title  \
id                                                                           
0        2020               MI              FT              Data Scientist   
1        2020               SE              FT  Machine Learning Scientist   
2        2020               SE              FT           Big Data Engineer   
3        2020               MI              FT        Product Data Analyst   
4        2020               SE              FT   Machine Learning Engineer   

    salary salary_currency employee_residence remote_ratio company_location  \
id                                                                            
0    70000             EUR                 DE            0               DE   
1   260000             USD                 JP            0               JP   
2    85000             GBP                 GB           50               GB   
3   -20000             USD                 HN            0               HN   
4   150000             USD                 US          ???               US   

   company_size  
id               
0             L  
1             S  
2             M  
3             S  
4             L

<div class="alert alert-block alert-info"><b>

## Aufgabe 1.0 </b> <br>
loc und iloc
</div>

<div class="alert alert-block alert-info"><b>Aufgabe 1.1 Mit iloc eine einzelne Zeile auswählen</b> <br>  
<div class="alert alert-block alert-warning">
Um die Daten zu reinigen müssen wir erste kenne lerne wie wir sie ändern können.<br>
Wir haben bereits df.head() und df.tail() kennengelernt.<br>
Nun wollen wir alle Werte einer bestimmten Zeile Auswählen. 
Dazu verwende wir den Befehl


 **DataFrame.iloc[Zeile]**


Schauen wir uns das ganze mal mit Probe Daten an.

</div>

</div>

In [126]:
#Vorgegeben
Test_data = pd.read_csv('Personen.csv',sep=';')
print(Test_data)
Test_data.iloc[0]

    vorname       nachname  alter
0  Christos    Tsinaraidis     23
1     Michi       Kaufmann     16
2     Ranji     Saseendran     25
3   Berkcan     Guezelocak     20
4    Tobias           Aign     20
5   Paulina  Geschwandtner     19


vorname        Christos
nachname    Tsinaraidis
alter                23
Name: 0, dtype: object

<div class="alert alert-block alert-info"><b>Aufgabe 1.1 Versuche nun die 4 Zeil aus dem Datafram anzuzeigen</b>

In [127]:
#Selber machen
df.iloc[3]

work_year                             2020
experience_level                        MI
employment_type                         FT
job_title             Product Data Analyst
salary                              -20000
salary_currency                        USD
employee_residence                      HN
remote_ratio                             0
company_location                        HN
company_size                             S
Name: 3, dtype: object

<div class="alert alert-block alert-info"><b> Aufgabe 1.2 mit loc eine Spalte auswählen </b>
<div class="alert alert-block alert-warning">
Wir kommen dem Ziel schon näher
Versuchen wir nun alle Werte in einer Spalte auszuwählen. <br>
Das Funktioniert mit dem Befehl:


**DataFrame.loc[:,'Spalten Name']**
</div>
</div>


In [128]:
#Vorgegeben
Test_data.loc[:,'vorname']

0    Christos
1       Michi
2       Ranji
3     Berkcan
4      Tobias
5     Paulina
Name: vorname, dtype: object

<div class="alert alert-block alert-warning"><b>Alternative</b> </br>
Wir können anstatt eine Spalten Namens auf eine liste mit Spaltenname übergeben <br>

**DataFrame.loc[:,['Spalten Name 1','Spalten Name 2']]**
</div>

In [129]:
Test_data.loc[:,['vorname','alter']]

vorname  alter
0  Christos     23
1     Michi     16
2     Ranji     25
3   Berkcan     20
4    Tobias     20
5   Paulina     19

<div class="alert alert-block alert-info"><b> Versuch jetzt die Spalten 'job_title', 'salary' und 'remote_ratio' auszuwählen </b></div>

In [130]:
# Selber machen 
df.loc[:,['job_title', 'salary', 'remote_ratio']]

job_title  salary remote_ratio
id                                                  
0                Data Scientist   70000            0
1    Machine Learning Scientist  260000            0
2             Big Data Engineer   85000           50
3          Product Data Analyst  -20000            0
4     Machine Learning Engineer  150000          ???
..                          ...     ...          ...
602               Data Engineer  154000          100
603               Data Engineer  126000          100
604                Data Analyst  129000            0
605                Data Analyst  150000          100
606                AI Scientist -200000          100

[607 rows x 3 columns]

<div class="alert alert-block alert-info"><b> Aufgabe 1.3 Kombiniere beide Methoden um einen einzelnen wert zu erhalten</b><br>

<div class="alert alert-block alert-warning">Mit dem Doppelpunk können wir alle Zeile auf einemal auswählen. Das wird im Spätern verlauf noch wichtig. <br>
Für die nächste Übung werden wir den doppelpunkt durch eine Zahl ersetzen.</div>

Last us den Doppelpunkt durch 0 ersetzen.
</div>

In [131]:
# Soll selber geschrieben werden
# vorergeben:
# Test_data.loc[]
Test_data.loc[0,'vorname']

'Christos'

<div class="alert alert-block alert-success"><b>Wird dir nur der Name Hristo angezeigt? </b><br>
Wenn ja hast du alles richtig gemacht. <br>


<div class="alert alert-block alert-info"><b> Aufgabe 1.4 Speicher den negativen wert in die Variabel 'negativerWert='</b>

Werfen wir doch mal einen Blick auf ersten 5 Zeilen. Siehst du den negativen wert in der Spalte 'salary'? <br>
Speicher den wert in der Variabel 'negativerWert='
Nutze print um zu prüfen ob du den richtigen hast.
</div>

In [132]:
df

work_year experience_level employment_type                   job_title  \
id                                                                            
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary salary_currency employee_residence remote_ratio company_location  \
id                                                                             
0     70000             EUR                 DE            0               DE   
1    260000             USD                 JP            0               JP   
2     85000             GBP                 GB           50               GB   
3    -20000             USD                 HN            0               HN   
4    150000             USD                 US          ???               US   
..      ...             ...                ...          ...              ...   
602  154000             USD                 US          100               US   
603  126000             USD                 US          100               US   
604  129000             USD                 US            0               US   
605  150000             USD                 US          100               US   
606 -200000             USD                 IN          100               US   

    company_size  
id                
0              L  
1              S  
2              M  
3              S  
4              L  
..           ...  
602            M  
603            M  
604            M  
605            M  
606            L  

[607 rows x 10 columns]

In [133]:
# Selber
# vorgegeben:
# negativerWert=
negativerWert=df.loc[3,'salary']
print(negativerWert)

-20000


<div class="alert alert-block alert-info"><b> Aufgabe 1.5 Eine Wert in der Tabelle ändern</b>

Wir haben nun eine Einzelne wert ausgehält.
<div class="alert alert-block alert-warning">
Loc und iloc sind quasi Positionsbeschreibungen. Wie könne mit der selben Method auch in DataFrame eine Wert ändern.<br>
</div>
Schau dir das beispiel an <br>
</div>



In [134]:
# Vorgeben
Test_data.loc[4,'vorname'] = 'Christopher'
Test_data

vorname       nachname  alter
0     Christos    Tsinaraidis     23
1        Michi       Kaufmann     16
2        Ranji     Saseendran     25
3      Berkcan     Guezelocak     20
4  Christopher           Aign     20
5      Paulina  Geschwandtner     19

<div class="alert alert-block alert-info"><b> Versuch mit der Selben Methode den Negativen Wert positiv umzurechnen  </b>

In [135]:
#Selber
df.loc[3,'salary'] = df.loc[3,'salary']*-1

In [136]:
df.head()

work_year experience_level employment_type                   job_title  \
id                                                                           
0        2020               MI              FT              Data Scientist   
1        2020               SE              FT  Machine Learning Scientist   
2        2020               SE              FT           Big Data Engineer   
3        2020               MI              FT        Product Data Analyst   
4        2020               SE              FT   Machine Learning Engineer   

    salary salary_currency employee_residence remote_ratio company_location  \
id                                                                            
0    70000             EUR                 DE            0               DE   
1   260000             USD                 JP            0               JP   
2    85000             GBP                 GB           50               GB   
3    20000             USD                 HN            0               HN   
4   150000             USD                 US          ???               US   

   company_size  
id               
0             L  
1             S  
2             M  
3             S  
4             L

<div class="alert alert-block alert-success"><b>  Wir habe jetzt einen Von 30 Werten gefunden und gereinigt. 😀</b>
Wir können mit dieser Methode jetzt natürlich auch die anderen reinigen. 
Wir wir können das aber auch den Computer machen lassen und dazu müssen wir lernen mehre Zeilen anhand eines Kriteriums auszuwählen. <br>
</div> 
<div class="alert alert-block alert-info"><b>

## Aufgabe 2.0 Masken </b> <br>
Wir müssen Lernen eine Maske zu erstellen
<div class="alert alert-block alert-warning">
Das geht ungefähr so

**mask = DataFram.loc[:,'Spalltenamen'] >= Wert**
</div>
Schauen wir und das ganze mal im Beispiel an
</div>

In [137]:
# Vorgegben
Test_maske= (Test_data.loc[:,'alter'] >= 20)
Test_data.loc[Test_maske,['vorname', 'nachname', 'alter']]

vorname     nachname  alter
0     Christos  Tsinaraidis     23
2        Ranji   Saseendran     25
3      Berkcan   Guezelocak     20
4  Christopher         Aign     20

<div class="alert alert-block alert-warning"> In der ersten Zeile haben wir ein Maske definiert. Dies Maske kann man an bei loc anstatt des doppelpunkten einsetzten. <br>
Zusammengefasst:
<li> Masken definieren welche Zeile angezeigt werde sollen</li>
</div>

<div class="alert alert-block alert-info">
Schreibe ein Maske um nur negative werte aus der Spalte Salary zu bekommen.
</div>

In [138]:
# Selber 
maske= df.loc[:,'salary'] <0
df.loc[maske]

work_year experience_level employment_type                   job_title  \
id                                                                            
34        2020               MI              FT                Data Analyst   
74        2021               EX              FT                Head of Data   
82        2021               MI              FT      Applied Data Scientist   
103       2021               MI              FT                Data Analyst   
115       2021               EN              FT  Machine Learning Scientist   
127       2021               MI              FT              Data Scientist   
156       2021               MI              FT              Data Scientist   
167       2021               EX              FT    Director of Data Science   
172       2021               EN              FT                Data Analyst   
173       2021               SE              FT    Principal Data Scientist   
195       2021               MI              FT              Data Scientist   
240       2021               SE              FT              Data Scientist   
243       2021               SE              FT              Data Scientist   
264       2021               MI              FT              Data Scientist   
279       2021               EN              FT       Business Data Analyst   
320       2022               SE              FT               Data Engineer   
373       2022               MI              FT               ETL Developer   
385       2022               SE              FT               Data Engineer   
422       2022               MI              FT        Data Science Manager   
438       2022               SE              FT   Machine Learning Engineer   
446       2022               SE              FT               Data Engineer   
463       2022               EN              FT              Data Scientist   
479       2022               MI              FT              Data Scientist   
498       2022               SE              FT          Research Scientist   
526       2022               MI              FT              Data Scientist   
540       2022               SE              FT                Data Analyst   
564       2022               SE              FT                Data Analyst   
600       2022               EN              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

      salary salary_currency employee_residence remote_ratio company_location  \
id                                                                              
34    -41000             EUR                 FR           50               FR   
74   -235000             USD                 US          100               US   
82    -68000             CAD                 GB           50               CA   
103   -62000             USD                 US            0               US   
115  -225000             USD                 US          100               US   
127  -700000             INR                 IN          ???               IN   
156  -160000             SGD                 SG          ???               IL   
167  -250000             USD                 US            0               US   
172   -60000             USD                 US          ???               US   
173  -235000             USD                 US          100               US   
195  -147000             USD                 US           50               US   
240  -130000             CAD                 CA          100               CA   
243  -165000             USD                 US          100               US   
264   -21600             EUR                 RS          100               DE   
279   -50000             EUR                 LU          100               LU   
320  -132320             USD                 US            0               US   
373   -50000             EUR                 GR            0               GR   
385  -132320    

<div class="alert alert-block alert-info"> <b> 2.2 Mit Masken Änderungen vornehmen </b>

Wir können nun Masken verwenden. Wir haben auch schon kenne gelernt wie wir Wert in der Tabelle änder können. <br>
Schau dir das Beispiel unten doch mal an.      
</div>

In [139]:
#voergegeben
Test_maske= (Test_data.loc[:,'alter'] >= 20)
Test_data.loc[Test_maske,'alter']= Test_data.loc[Test_maske,'alter']+43
Test_data


vorname       nachname  alter
0     Christos    Tsinaraidis     66
1        Michi       Kaufmann     16
2        Ranji     Saseendran     68
3      Berkcan     Guezelocak     63
4  Christopher           Aign     63
5      Paulina  Geschwandtner     19

<div class="alert alert-block alert-success">Und Schon sind alle ausgewählten im Renten Alter. </div>


<div class="alert alert-block alert-info"> <b> Nutze das selbe Prinzip um alle negativen Werte durch positive Werte zu ersetzen.</b>
</div>

In [140]:
# selber
df.loc[maske,'salary']= df.loc[maske,'salary']*-1
df

work_year experience_level employment_type                   job_title  \
id                                                                            
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary salary_currency employee_residence remote_ratio company_location  \
id                                                                             
0     70000             EUR                 DE            0               DE   
1    260000             USD                 JP            0               JP   
2     85000             GBP                 GB           50               GB   
3     20000             USD                 HN            0               HN   
4    150000             USD                 US          ???               US   
..      ...             ...                ...          ...              ...   
602  154000             USD                 US          100               US   
603  126000             USD                 US          100               US   
604  129000             USD                 US            0               US   
605  150000             USD                 US          100               US   
606  200000             USD                 IN          100               US   

    company_size  
id                
0              L  
1              S  
2              M  
3              S  
4              L  
..           ...  
602            M  
603            M  
604            M  
605            M  
606            L  

[607 rows x 10 columns]

<div class="alert alert-block alert-success">Prüfen wir doch mal ob du alles richtig gemacht hast.
Führe die Code Zelle unten aus </div>

In [141]:

if pd.util.hash_pandas_object(df['salary']).sum() ==-7433296294440947629:
    print("Super du hast all negativen Werte korrigiert")
else:
    raise Warning('leider stimmt da was nicht')

Super du hast all negativen Werte korrigiert


<div class="alert alert-block alert-info"> <b>

## Aufgabe 3.0 Datentypen </b>

 Zum Daten reinigen gehört aber nicht nur Falsche Daten zu richten. Sonder auch die Daten in ihre optimale from zu bringen <br>
Hir werden jetzt Datentypen wichtig. </div>
 <div class="alert alert-block alert-warning"> Schaun wir uns doch mal die Datentypen der einzelne spalten an. Das geht mit dem Befehl: <br>
 
**DataFrame.dtypes** 
 
</div>

In [142]:
# selber
df.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
employee_residence    object
remote_ratio          object
company_location      object
company_size          object
dtype: object

<div class="alert alert-block alert-warning">

Bestimmt ist dir aufgefallen das der Spaltentyp der meisten Spalten "object" ist. 
Ein spalte kann aus zwei gründen als object angzeigt werden.
1. Pander erkennt nicht den richtigen Datentyp.
Das ist bei uns der Fall
2. Der Datentyp ist nicht einheitlich.
Das ist der komplizierter Fall. Es ist ein hinwies das der Datensatz verunreinigt ist.<br><br>


Wichtig ist : 

**Verschiedene Datentypen haben unterschiedlich viel Speicher bedarf**
</div>


<div class="alert alert-block alert-info"> <b> Aufgabe 3.2 Speicherbedarf </b><br>
Schauen wir uns doch mal die Größe unsers Dataframe mal in. 
<div class="alert alert-block alert-warning">

Das get mit dem Befehl <br>
**DataFrame.info()**
</div>

</div>

In [143]:
#selber
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 607 entries, 0 to 606
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           607 non-null    int64 
 1   experience_level    607 non-null    object
 2   employment_type     607 non-null    object
 3   job_title           607 non-null    object
 4   salary              607 non-null    int64 
 5   salary_currency     607 non-null    object
 6   employee_residence  607 non-null    object
 7   remote_ratio        607 non-null    object
 8   company_location    607 non-null    object
 9   company_size        607 non-null    object
dtypes: int64(2), object(8)
memory usage: 68.3+ KB


<div class="alert alert-block alert-warning">Wie wie du sicher bemerkt gibt 'info' neben dem Speicherbedarf auch die Datentypen an. Der Nachteil ist aber, das endgegensatz zu 'dtypes', die Datentypen nicht so leicht als "list" gespeichert werden können.
</div>

<div class="alert alert-block alert-info"> <b>Aufgabe 3.3 Datentypen ändern </b> <br>

Es kann viele gründe geben den Datentypen zu ändern. Unser ist jedoch den Speicherbedarf zu reduzieren.
<div class="alert alert-block alert-warning">

Das ganze funktioniert mit dem Befehl: <br>
**DataFrame['column'] = DataFrame['column'].astype('datentyp')**

</div>
Schau dir doch mal das beispiel an
</div>

In [144]:
# Vorgegeben
Test_data

vorname       nachname  alter
0     Christos    Tsinaraidis     66
1        Michi       Kaufmann     16
2        Ranji     Saseendran     68
3      Berkcan     Guezelocak     63
4  Christopher           Aign     63
5      Paulina  Geschwandtner     19

In [145]:
#Vorgegebenen
print(Test_data.info(verbose=False))
Test_data['alter']= Test_data['alter'].astype('int8')
print(Test_data.info(verbose=False))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Columns: 3 entries, vorname to alter
dtypes: int64(1), object(2)
memory usage: 276.0+ bytes
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Columns: 3 entries, vorname to alter
dtypes: int8(1), object(2)
memory usage: 234.0+ bytes
None


<div class="alert alert-block alert-warning">
In diesem Fall reicht es Komplet aus das Alter in einem Int8 zu speichern(Int8 = byte). Die Tabelle unten gibt Aufschluss über die Zahlenbereiche.<br><br>
Zusammengefasst: Schränke den Zahlenbereich soweit wie möglich ein um Speicher zu sparen.
</div>




<div class="alert alert-block alert-info"> <b> Aufgabe 3.3.1 Änder den Datentyp der Spalte 'work_year' zu einem Passenden Type </b>


|Type|Number of bits|Min. value|Max. value|
|-:| -: | -: | -: |
|int8|8 bit|–128|127|
|int16|16 bit|–32768|32767|
|int32|32 bit|–2147483648|2147483647|
|int64|64 bit|–2^63|2^63 - 1|
</div>

In [146]:
# Selber
df['work_year']=df['work_year'].astype('Int16')
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 607 entries, 0 to 606
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           607 non-null    Int16 
 1   experience_level    607 non-null    object
 2   employment_type     607 non-null    object
 3   job_title           607 non-null    object
 4   salary              607 non-null    int64 
 5   salary_currency     607 non-null    object
 6   employee_residence  607 non-null    object
 7   remote_ratio        607 non-null    object
 8   company_location    607 non-null    object
 9   company_size        607 non-null    object
dtypes: Int16(1), int64(1), object(8)
memory usage: 65.4+ KB


<div class="alert alert-block alert-info"> <b>Aufgabe 3.4 Effiziente Speicherung von Strings </b> <br>
Es gibt auch ein Möglichkeit um Strings effizient zu speichern. Das ganze ist aber etwas situationsabhängig. Der Datentyp nennt sich category. <br>
Dabei wird für jeden einzigartigen string eine Category erstellt. Heist das Ganze macht nur dan sin wenn wir ein überschaubare menge an einzigartigen Strings in einer Spalte haben

<div class="alert alert-block alert-info"> <b> Aufgabe 3.4.1 unique Data </b> <br>

Wir müssen herausfinden wie viel verscheiden einzigartige stings es in unser Tabelle gibt.<br>
<div class="alert alert-block alert-warning"> 

Das funktioniert über den Befehle:<br>
**df['column'].unique()**
</div>
Finde heraus wie viel einzigartige Werte es in der Spalte "company_location" gibt
</div>

In [147]:
#selber
len(df['company_location'].unique())


50

<div class="alert alert-block alert-info"> <b> Aufgabe Aufgabe 3.4.2 änder alle Spalten mit dem Datentyp object zum Typ 'category' </b> <br>
tip: erstelle erst ein list mit allen Spalten. Zeige danach die Speicher Größe an
</div>

In [148]:
# selber
for col in df.columns:
    # df.loc[:, col] = df.loc[:, col].astype('category')
    if df[col].dtype == "object":
        df[col]=df[col].astype('category')
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 607 entries, 0 to 606
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   work_year           607 non-null    Int16   
 1   experience_level    607 non-null    category
 2   employment_type     607 non-null    category
 3   job_title           607 non-null    category
 4   salary              607 non-null    int64   
 5   salary_currency     607 non-null    category
 6   employee_residence  607 non-null    category
 7   remote_ratio        607 non-null    category
 8   company_location    607 non-null    category
 9   company_size        607 non-null    category
dtypes: Int16(1), category(8), int64(1)
memory usage: 41.0 KB


<div class="alert alert-block alert-info"> <b>

## Aufgabe 4.0 Optionale Aufgabe </b>

<b> 4.1 (Optional) sorge dafür das in der Spalte 'remote_ratio' anstatt fragezeichen numpy np.nan werte stehen    </b> <br>
Wenn in unser Spalten werte auftauchen welche unrealistisch oder erkennbar falsch sind sollte man diese spalten entweder komplett entferne oder die fehlerhaften Wert so kennzeichen das sie bei späteren Operationen nicht auffallen.

<b> 4.1.a (Optional) Änder alle "???" Einträge zu 'np.nan' </b> <br>
Da wir den ursprünglichen wert von "remote_ratio" nicht kennen bleibt uns nichts anders übrig als diese zu ignoriere oder ganz zu löschen.
Fürs erst wollen wir den Wert ignorieren 
Wir könne den np.nan wert nutzen um sicher zustellen das die werte in Kalkulationen nicht mehr verwendet werden.
Nutze ein Maske um alle betroffene Zeilen zu 'np.nan' zu ändern
</div>

In [149]:
# selber
maske= df.loc[:,'remote_ratio'] == "???"
df.loc[maske,'remote_ratio']= np.nan
df.head()



work_year experience_level employment_type                   job_title  \
id                                                                           
0        2020               MI              FT              Data Scientist   
1        2020               SE              FT  Machine Learning Scientist   
2        2020               SE              FT           Big Data Engineer   
3        2020               MI              FT        Product Data Analyst   
4        2020               SE              FT   Machine Learning Engineer   

    salary salary_currency employee_residence remote_ratio company_location  \
id                                                                            
0    70000             EUR                 DE            0               DE   
1   260000             USD                 JP            0               JP   
2    85000             GBP                 GB           50               GB   
3    20000             USD                 HN            0               HN   
4   150000             USD                 US          NaN               US   

   company_size  
id               
0             L  
1             S  
2             M  
3             S  
4             L

<div class="alert alert-block alert-info"> <b> 4.1.b (Optional) Ändere den Spalten typen zu float </b> <br>
Die falschen werte sind ausgeschlossen aber das reicht noch nicht ganz. Bestimmte Funktionen wir .mean() setzten Numerisch Spaltentypen voraus. Pandas erlaubt aber keine nan werte in 'int' spalten. Aber in "float" spalten sind nan werte erlaubt.<br>
Folge änder die Spalte 'remote_ration' zum typ float
Tip: wenn der aktuelle Spalten typ category ist änder den typ erst zu string und dan zu float
</div>

In [150]:
#selber
df['remote_ratio'] = df['remote_ratio'].astype('string')
df['remote_ratio'] = df['remote_ratio'].astype('float')

<div class="alert alert-block alert-info"> <b> 4.1.c (Optional) berechne den die durchschnittliche Home office quote </b> <br>
Wenn alle Daten gereinigt sind können wir den durchschnitt berechnen.
Tip nutze den Befehl '.mean()' 


In [151]:
# selber
df['remote_ratio'].mean()

71.10726643598616

<div class="alert alert-block alert-info"> <b> 4.2 (Optional) "???" direkt als Nan Werte Kennzeichnen </b> <br>

Es gibt auch die Möglichkeit sich das suche nach einen einheitliche Fehlerwerten zu sparen. 
Übergebe bei dem Befehl 'pd.read_csv' dem optionalen Parameter 'na_values' mit dem Wert '???'

In [153]:
# Selber
pd.read_csv('ds_salaries_real01.csv',index_col='id',na_values='???')


work_year experience_level employment_type                   job_title  \
id                                                                            
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary salary_currency employee_residence  remote_ratio company_location  \
id                                                                              
0     70000             EUR                 DE           0.0               DE   
1    260000             USD                 JP           0.0               JP   
2     85000             GBP                 GB          50.0               GB   
3    -20000             USD                 HN           0.0               HN   
4    150000             USD                 US           NaN               US   
..      ...             ...                ...           ...              ...   
602  154000             USD                 US         100.0               US   
603  126000             USD                 US         100.0               US   
604  129000             USD                 US           0.0               US   
605  150000             USD                 US         100.0               US   
606 -200000             USD                 IN         100.0               US   

    company_size  
id                
0              L  
1              S  
2              M  
3              S  
4              L  
..           ...  
602            M  
603            M  
604            M  
605            M  
606            L  

[607 rows x 10 columns]